mlrun/model.py

```diff
@@ -56,22 +60,115 @@ def _verify_dict(param, name, new_type=None):
+        First we run over the fields and serialize them, then we run over the fields and enrich them.
+        :param exclude: A list of fields to exclude from the dictionary. If not provided, the default value is an
+         empty list and the object's `_fields_to_exclude_for_serialization` attribute is appended to it.
+        :param strip:  If True, the object's `_fields_to_strip` attribute is appended to the exclude list.
+        :return: A dictionary representation of the object.
+        """
         struct = {}
-        fields = fields or self._dict_fields
-        if not fields:
-            fields = list(inspect.signature(self.__init__).parameters.keys())
-        for t in fields:
-            if not exclude or t not in exclude:
-                val = getattr(self, t, None)
-                if val is not None and not (isinstance(val, dict) and not val):
-                    if hasattr(val, "to_dict"):
-                        val = val.to_dict()
-                        if val:
-                            struct[t] = val
-                    else:
-                        struct[t] = val
+
+        fields = self._resolve_fields(fields)
+        fields_to_exclude = exclude or []
+        if strip:
+            fields_to_exclude += self._fields_to_strip
+
+        fields_to_save = (
+            set(fields)
+            - set(fields_to_exclude)
+            - set(self._fields_to_exclude_for_serialization)
+            - set(self._fields_to_exclude_for_enrichment)
+        )
+
+        for field in fields_to_save:
+            field_value = getattr(self, field, None)
+            if self._is_valid_field_value(field_value):
+                if hasattr(field_value, "to_dict"):
+                    field_value = field_value.to_dict(strip=strip)
+                    if self._is_valid_field_value(field_value):
+                        struct[field] = field_value
+                else:
+                    struct[field] = field_value
+
+        # if field is in field_to_exclude_for_serialization is also in the fields_to_exclude list,
+        # then no need to iterate over it
+        fields_to_serialize = list(
+            set(self._fields_to_exclude_for_serialization) - set(fields_to_exclude)
+        )
+        self._resolve_field_by_method(
+            struct, self._serialize_field, fields_to_serialize, strip
+        )
+
+        # if field is in _fields_to_exclude_for_enrichment is also in the fields_to_exclude list, it means that the user
+        # explicitly excluded it from the serialization, so we won't enrich it.
+        fields_to_enrich = list(
+            set(self._fields_to_exclude_for_enrichment) - set(fields_to_exclude)
+        )
+        self._resolve_field_by_method(
+            struct, self._enrich_field, fields_to_enrich, strip
+        )
+
+        self._apply_enrichment_before_to_dict_completion(struct, strip=strip)
+        return struct
+
+    def _resolve_fields(self, fields: list = None) -> list:
```

Comment:
The name is too general (and `resolve` is used in many places) - suggest to change this to `_resolve_initial_to_dict_fields` or something more indicative.